internal/api/v1beta1/app_types.go

```diff
@@ -665,3 +657,200 @@ func (t Target) IsDeployment() bool {
+	// values: CanaryStarted, CanaryFinished
+	// errored values: CanaryNotActiveEvent, CanaryNoDeployments, CanaryNoScheduledSteps
+	Name string
+	// Description states what is the outcome of this event
+	Description string
+}
+
+func newCanaryEvent(app *App, event string, desc string) CanaryEvent {
+	var version DeploymentVersion
+	if len(app.Spec.Deployments) > 0 {
+		version = app.Spec.Deployments[len(app.Spec.Deployments)-1].Version
+	}
+
+	return CanaryEvent{
+		AppName:           app.Name,
+		DeploymentVersion: int(version),
+		Name:              event,
+		Description:       desc,
+	}
+}
+
+// Message is message for k8s event
+func (c CanaryEvent) Message() string {
+	return fmt.Sprintf("%s - Canary for app %s | version %d - %s", c.Name, c.AppName, c.DeploymentVersion, c.Description)
+}
+
+// CanaryEventFromMessage creates CanaryEvent from given message
+// NOTE as description can contain spaces, it won't be perfectly populated for some events
+func CanaryEventFromMessage(msg string) (*CanaryEvent, error) {
+	event := CanaryEvent{}
+
+	_, err := fmt.Sscanf(msg, "%s - Canary for app %s | version %d - %s", &event.Name, &event.AppName, &event.DeploymentVersion, &event.Description)
+	if err != nil {
+		return nil, fmt.Errorf(`unable to parse CanaryEvent: %s, err: %v`, msg, err)
+	}
+
+	return &event, nil
+}
+
+type CanaryNextStepEvent struct {
+	Event CanaryEvent
+
+	Step          int
+	VersionSource int
+	VersionDest   int
+	WeightSource  uint8
+	WeightDest    uint8
+}
+
+func newCanaryNextStepEvent(app *App) CanaryNextStepEvent {
+	return CanaryNextStepEvent{
+		Event: newCanaryEvent(app, CanaryNextStep, CanaryNextStepDesc),
+
+		Step:          app.Spec.Canary.CurrentStep,
+		VersionSource: int(app.Spec.Deployments[0].Version),
+		VersionDest:   int(app.Spec.Deployments[1].Version),
+		WeightSource:  app.Spec.Deployments[0].RoutingSettings.Weight,
+		WeightDest:    app.Spec.Deployments[1].RoutingSettings.Weight,
+	}
+}
+
+func (c CanaryNextStepEvent) Message() string {
+	return fmt.Sprintf("%s: Step: %d | Source version: %d | Dest version: %d | Source weight: %d | Dest weight: %d", c.Event.Message(), c.Step, c.VersionSource, c.VersionDest, c.WeightSource, c.WeightDest)
+}
+
+// CanaryNextStepEventFromString creates CanaryNextStepEvent from given message
+// NOTE as description can contain spaces, it won't be perfectly populated for some events
+func CanaryNextStepEventFromString(msg string) (*CanaryNextStepEvent, error) {
+	split := strings.SplitAfterN(msg, ":", 2)
+	if len(split) < 2 {
+		return nil, errors.New("invalid message format")
+	}
+
+	canaryEvent, err := CanaryEventFromMessage(split[0])
+	if err != nil {
+		return nil, err
+	}
+
+	event := CanaryNextStepEvent{
+		Event: *canaryEvent,
+	}
+
+	_, err = fmt.Sscanf(split[1], " Step: %d | Source version: %d | Dest version: %d | Source weight: %d | Dest weight: %d", &event.Step, &event.VersionSource, &event.VersionDest, &event.WeightSource, &event.WeightDest)
```

Comment:
With @koncar we agreed on using an event's annotations to store this information instead of parsing it